src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompileRunCommand.cs

```diff
@@ -0,0 +1,193 @@
+                updatedAssembly = rewriter.Rewrite(updatedAssembly);
+            }
+
+            using (var stream = File.OpenWrite(Path.ChangeExtension(assemblyPath, ModifiedAssemblyExtension)))
+            {
+                PeWriter.WritePeToStream(updatedAssembly, host, stream);
+            }
+        }
+
+        private PrecompilationResult CompileViews(IFileProvider fileProvider)
+        {
+            var precompilationResult = new PrecompilationResult();
+            foreach (var fileInfo in FileProviderUtilities.GetRazorFiles(fileProvider))
+            {
+                CompileView(precompilationResult, fileInfo);
+            }
+
+            return precompilationResult;
+        }
+
+        private void CompileView(PrecompilationResult result, RelativeFileInfo fileInfo)
+        {
+            GeneratorResults generatorResults;
+            using (var fileStream = fileInfo.FileInfo.CreateReadStream())
+            {
+                generatorResults = Host.GenerateCode(fileInfo.RelativePath, fileStream);
+            }
+
+            if (!generatorResults.Success)
+            {
+                result.RazorErrors.AddRange(generatorResults.ParserErrors);
+                return;
+            }
+
+            var compileOutputs = new CompileOutputs(fileInfo.RelativePath, GeneratePdbOption.HasValue());
+            var emitResult = CompilationService.EmitAssembly(
+                PrecompiledAssemblyPrefix + Guid.NewGuid(),
+                generatorResults.GeneratedCode,
+                compileOutputs.AssemblyStream,
+                compileOutputs.PdbStream);
+
+            if (!emitResult.Success)
+            {
+                result.RoslynErrors.AddRange(emitResult.Diagnostics);
+                compileOutputs.Dispose();
+            }
+            else
+            {
+                result.CompileOutputs.Add(compileOutputs);
+            }
+        }
+
+        private IServiceProvider ConfigureDefaultServices()
+        {
+            var services = new ServiceCollection();
+
+            var applicationName = Path.GetFileName(ProjectPath.TrimEnd('/'));
+            var contentRoot = ContentRootOption.HasValue()
+                ? ContentRootOption.Value()
+                : Directory.GetCurrentDirectory();
+            var hostingEnvironment = new HostingEnvironment
+            {
+                ApplicationName = applicationName,
+                WebRootFileProvider = new PhysicalFileProvider(ProjectPath),
+                ContentRootFileProvider = new PhysicalFileProvider(contentRoot),
+                ContentRootPath = contentRoot,
+            };
+            var diagnosticSource = new DiagnosticListener("Microsoft.AspNetCore");
+
+            services
+                .AddSingleton<IHostingEnvironment>(hostingEnvironment)
+                .AddSingleton<DiagnosticSource>(diagnosticSource)
+                .AddLogging()
+                .AddSingleton<ObjectPoolProvider, DefaultObjectPoolProvider>();
```

Comment:
It's not that we're terrible, it's that there's no way to "register hosting services" that we can call, and our `AddXyz` aren't self-sufficient.